### PR TITLE
Avoid captures for AnyAsync

### PR DESCRIPTION
**Pull request by sharwell**
These captures showed up as a few GB during a long-running high CPU operation measured locally.

### CODE REVIEW

**Review by davidwengier — APPROVED** [review]

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 1]
```diff
@@ -422,6 +422,22 @@ public static async Task<bool> AnyAsync<T>(this ImmutableArray<T> array, Func<T,
             return false;
         }
 
+        public static async Task<bool> AnyAsync<T, TArg>(this ImmutableArray<T> array, Func<T, TArg, Task<bool>> predicateAsync, TArg arg)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:425]: Value task?

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 2]
```diff
@@ -422,6 +422,22 @@ public static async Task<bool> AnyAsync<T>(this ImmutableArray<T> array, Func<T,
...
+        {
+            int n = array.Length;
+            for (int i = 0; i < n; i++)
+            {
+                var a = array[i];
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:430]: Why not basic for each?  It's non allocating for immutable array right?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs [hunk 3]
```diff
@@ -54,8 +54,8 @@ private static bool IsAccessible(ISymbol symbol)
                 var namespace2Count = namespace2.ConstituentNamespaces.Length;
                 if (namespace1Count != namespace2Count)
                 {
-                    if ((namespace1Count > 1 && await namespace1.ConstituentNamespaces.AnyAsync(n => NamespaceSymbolsMatchAsync(solution, n, namespace2, cancellationToken)).ConfigureAwait(false)) ||
-                        (namespace2Count > 1 && await namespace2.ConstituentNamespaces.AnyAsync(n2 => NamespaceSymbolsMatchAsync(solution, namespace1, n2, cancellationToken)).ConfigureAwait(false)))
+                    if ((namespace1Count > 1 && await namespace1.ConstituentNamespaces.AnyAsync(static (n, arg) => NamespaceSymbolsMatchAsync(arg.solution, n, arg.namespace2, arg.cancellationToken), (solution, namespace2, cancellationToken)).ConfigureAwait(false)) ||
```
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs:57]: Can the other extension method be deleted in favor of the new one?

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:425]: It probably would be if this were in the runtime, but for us it makes no difference. This form matches the methods already in this file.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs:57]: I'm not sure. The other one is perfectly valid so I didn't look outside the ones that appeared in the trace.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs:430]: I'm not sure why everything in this file is written this way, but I prefer not to change it for just one case without cause, especially as part of a performance PR.

**Review by cston — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Helpers.cs:57]: I gave a look and it seems this PR replaced all usages of the old extension method.

https://grep.app/search?q=.AnyAsync&filter[repo][0]=dotnet/roslyn

It had only 3 usages which you replaced in the PR. I think it should be fine either way.

**Pull request merged**